### ISSUE TITLE
Is support for k-fold cross-validation on the roadmap?

### ISSUE THREAD

**Issue by ianhellstrom** [tensorflow/tfx#121]
Right now, the training/evaluation split is managed by TFX internally. For making sure model parameters are set appropriately and to avoid having a 'lucky' split affecting the model quality, k-fold cross-validation be useful to have baked into TFX. Is this on the roadmap?

Moreover, will there be support for train/evaluation/validation splits, which is especially useful when tuning hyperparameters of deep networks. The evaluation data may bleed into the training procedure by tweaking the hyperparameters, so that a pure validation data set would be ideal: one the model does not see during training or hyperparameter tuning. Is this on the TFX's radar?

**Comment by 1025KB** [comment]
Currently for ExampleGen, you are able to customize the input and output split [doc](Update chicago_taxi to keep up with TFX 0.13 release), but downstream components only support train and eval split for now

We don't have a offical support yet, but you can customize the Trainer component for the k-fold and train/evaluation/validation splits, you can think tfx eval split is the validation split you mentioned, and tfx train split is the train/evaluation you mentioned, in custom Trainer component, convert the train split as train/evaluation and do k-fold for tuning the model

I will keep this PR as a future request

**Comment by gowthamkpr** [comment]
I am gonna close this issue as it has been resolved. Please add additional comments and we can open this issue again. Thanks!

**Comment by lelayf** [comment]
It is not clear to me how this issue has been resolved. Can someone advise?

**Comment by 1025KB** [comment]
Hi, Currently [custom split](https://github.com/tensorflow/tfx/blob/master/docs/guide/examplegen.md#custom-inputoutput-split) is fully supported for examplegen and its downstream, you should be able to utilize the splits in GenericTrainer's run_fn based on the needs

**Comment by lelayf** [comment]
I consume a balanced dataset (50% positive/50% negative labels for binary classification), what I need is the equivalent of scikit-learn's StratifiedKFold, which ensures every split remains balanced. At the moment this does not seem possible and my splits are heavily imbalanced after splitting. Also, I don't think the partition_feature_name would help this use case.

**Comment by 1025KB** [comment]
I see, our split is just random split, so it might because the dataset is relatively small thus the results are not perfectly balanced as input.

Tensorflow [dataset has an api](https://stackoverflow.com/questions/39748660/how-to-perform-k-fold-cross-validation-with-tensorflow) to split data, but I'm not sure if that will guarantee the balance. If that works, you can just pass a single train split to GenericTrainer, and in the run_fn, split the dataset and do the cross validation.
